### ISSUE TITLE
Quarto ignores `output:asis` and throws `cell-output-display` div

### ISSUE THREAD

**Issue by AlbertRapp** [quarto-dev/quarto-cli#7600]
### Bug description

Quarto seems to ignore the `output:asis` command for some reason and throws the `cell-output-display` anyway.

### Steps to reproduce

````qmd


:::: {style="display:grid;grid-template-columns:repeat(3, 1fr);gap:20px;"}

```{r}
#| echo: false
#| output: asis

htmltools::tagList(
    purrr::map(LETTERS, ~htmltools::div(.))
)
```

::::

````

### Expected behavior

The letters should be aligned on a grid with three columns.
![image](https://github.com/quarto-dev/quarto-cli/assets/65388595/a4cf1c7c-667b-493f-a6c8-9aa4cb6e175a)

### Actual behavior

All letters get placed below each other into the first column of the grid (because all the letters are in a `div` of class `cell-output-display` 
![image](https://github.com/quarto-dev/quarto-cli/assets/65388595/681de5aa-a2d4-413f-9c0f-7f087e84997a)

### Your environment

 version  R version 4.3.2 (2023-10-31)
 os       Ubuntu 22.04.3 LTS
 system   x86_64, linux-gnu
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Berlin
 date     2023-11-15
 rstudio  2023.12.0-daily+237 Ocean Storm (desktop)
 pandoc   NA

### Quarto check output

```
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.3.450
      Path: /opt/quarto/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.10.12
      Path: /home/albert/.virtualenvs/r-reticulate/bin/python3
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 4.3.2
      Path: /usr/lib/R
      LibPaths:
        - /home/albert/R/x86_64-pc-linux-gnu-library/4.3
        - /usr/local/lib/R/site-library
        - /usr/lib/R/site-library
        - /usr/lib/R/library
      knitr: 1.44
      rmarkdown: 2.25

[✓] Checking Knitr engine render......OK
```

**Comment by cscheid** [comment]
This isn't a bug. `cell-output-display` is not an optional div for quarto. `output: asis` changes the behavior of what goes inside the cell-output-display divs, but not the presence of that div itself.

Since you're using `output: asis`, you can emit any DOM you want on the inside, including a new div with other classes that you can target with CSS.

**Comment by mcanouil** [comment]
FYI, there are several examples in the discussions using `knitr::knit_child` & co
See <https://github.com/quarto-dev/quarto-cli/discussions?discussions_q=knit_child>.

**Comment by AlbertRapp** [comment]
> `cell-output-display` is not an optional div for quarto. `output: asis` changes the behavior of what goes inside the `cell-output-display` divs, but not the presence of that div itself.

I'm very surprised by this. [The docs state](https://quarto.org/docs/reference/cells/cells-knitr.html) that *"asis: Treat output as raw markdown with no enclosing containers."*

And this works perfectly fine, when my output is not a taglist. For example, here's a `list` output.

## Example that works as expected

````qmd
:::: {style="display:grid;grid-template-columns:repeat(3, 1fr);gap:20px;"}

```{r}
#| echo: false
#| output: asis
list(
    'This has no surrounding container',
    'This has no surrounding container',
    'This has no surrounding container',
    'This has no surrounding container'
)
```

::::
````

This gives me the desired result: 

![image](https://github.com/quarto-dev/quarto-cli/assets/65388595/b4dddec6-9566-4ddf-9075-a95ace79f7b6)

And if I look into the source code of the generated html file, there's no surrounding container with class `cell-output-display`. The surrounding `div` that is there is the one that I have included in the Quarto file manually.

![image](https://github.com/quarto-dev/quarto-cli/assets/65388595/e8be2368-1835-449f-b974-4533ca4cbc5a)


## Modifying `tagList()` example so that it triggers similar behavior

Also, the example from @mcanouil works perfectly fine (Thank you!) but I suspect that it works because it outputs not a `tagList()` but something that is generated from some other code. Still, the result is *not* enclosed in a surrounding `div`. 

Finally, the desired behavior from my `tagList()` code chunk can be enforced by including something which is **not** a tagList():

````qmd
:::: {style="display:grid;grid-template-columns:repeat(3, 1fr);gap:20px;"}

```{r}
#| echo: false
#| output: asis
3 # Add some non-taglist output
htmltools::tagList(
    purrr::map(LETTERS, ~htmltools::div(.))
)
```

::::

````

This will throw in a `div` that seem to get wrapped around the `htmltools::tagList()` but only as plain text. 

![image](https://github.com/quarto-dev/quarto-cli/assets/65388595/cf01891a-8ea9-4f65-981a-43b3ee61afba)


So long story short, there seems to be a different behavior from `output:asis` with `tagList()` in comparison to what `output:asis` does with e.g. regular `list()` (see example above). It would be desirable if one could get the desired output without the extra steps of ` |> as.character() |> cat()` (which works, does not need the enclosing `::::` and which I document here in case someone stumbles on this) :

````qmd
:::: {style="display:grid;grid-template-columns:repeat(3, 1fr);gap:20px;"}

```{r}
#| echo: false
#| output: asis

# This works without
htmltools::tagList(
    purrr::map(LETTERS, ~htmltools::div(.))
) |> as.character() |> cat()
```

::::

````

![image](https://github.com/quarto-dev/quarto-cli/assets/65388595/881ff5ba-88ea-4c04-9fce-54149a899031)

**Comment by mcanouil** [comment]
> And this works perfectly fine, when my output is not a taglist. For example, here's a list output.

I believe it is knitr that has a print method for the fonction which makes the current behaviour.

@cderv will have the best explanation for this 😏

**Comment by AlbertRapp** [comment]
Ahh, so it is either `htmltools`  or `knitr` that wraps the `tagList()` output into another `div` with class `cell-output-display` so that this behaviour should be mentioned in those repos and not here? I originally thought Quarto did that. My bad.

**Comment by mcanouil** [comment]
No, the thing is what's being emitted. by `htmltools`/`knitr`.
It does not emit something that is plain markdown.
It outputs HTML code which is embedded in raw block to ensure it will be read "as is", this triggers Quarto to surround the output as it is not markdown, it is a raw code block.

Look at the intermediate markdown from the code below:

````qmd
---
title: "Quarto Playground"
format: html
keep-md: true
---

## Original

:::: {style="display:grid;grid-template-columns:repeat(3, 1fr);gap:20px;"}

```{r}
#| echo: false
#| output: asis

htmltools::tagList(
    purrr::map(LETTERS, ~htmltools::div(.))
)
```

::::

## Corrected

```{r}
#| echo: false
#| output: asis
cat(
  '::: {style="display:grid;grid-template-columns:repeat(3, 1fr);gap:20px;"}',
  as.character(htmltools::tagList(purrr::map(LETTERS, ~htmltools::div(.)))),
  ":::",
  sep = "\n"
)
```
````

**Comment by cderv** [comment]
Thanks for asking the question. 

This is related to this discussion

* https://github.com/quarto-dev/quarto-cli/discussions/6874

In short, with **knitr** engine, any output of class `knit_asis` will be wrapped into this `.cell-output-display` div. Only thing that `output: asis` do currently is remove the outter `.cell `div. 

However, for Jupyter it does not work the same. So it still makes me wonder if we doing the right thing here. Let's look into this. 

## `output: asis` does not always add a div

<table>
<tr>
 <td>
 <td>Jupyter
 <td>Knitr
<tr>
 <td>.qmd
 <td>


````markdown
---
format: html
keep-md: true
---

```{python}
#| echo: false
#| output: asis
print("Hello **Markdown**")
```

````
 <td>

````markdown
---
format: html
keep-md: true
---

```{r}
#| echo: true
#| output: asis
cat("Hello **Markdown**")
```

````

<tr>
 <td>.md
 <td>

````markdown
---
format: html
keep-md: true
---


Hello **Markdown**

````
 <td>

````markdown
---
format: html
keep-md: true
---


Hello **Markdown**



````

</table>

So here, it happens because something else. Probably the tagList that is supposed to produced raw HTML code

<table>
<tr>
 <td>
 <td>Jupyter
 <td>Knitr
<tr>
 <td>.qmd
 <td>

````markdown
---
format: html
keep-md: true
---

```{python}
#| echo: false
#| output: asis
from IPython.display import HTML
HTML("<div>Hello</div>")
```

````
 <td>


````markdown
---
format: html
keep-md: true
---

```{r}
#| echo: false
#| output: asis
htmltools::HTML("<div>Hello</div>")
```


````

<tr>
 <td>.md
 <td>

````markdown
---
format: html
keep-md: true
---

```{=html}
<div>Hello</div>
```


````
 <td>

---
format: html
keep-md: true
---


::: {.cell-output-display}


```{=html}
<div>Hello</div>
```


:::


</table>



We do add the div in the **knitr** case, but we do not in the **Jupyter** engine case. 

> This isn't a bug. cell-output-display is not an optional div for quarto. output: asis changes the behavior of what goes inside the cell-output-display divs, but not the presence of that div itself.

@cscheid See my example above. Does this means we don't do the right thing for Jupyter ? 

Our doc also says (at https://quarto.org/docs/computations/execution-options.html#raw-output)

> When output: asis is specified none of Quarto’s standard enclosing divs will be included. 

This is the case for simple content, but not for specific `knit_asis` content. Do we miss something in **knitr** to mark content a specific way for quarto to know ? 

I don't fully understand why we would add it. In Jupyter, we do not add this div no matter what when `output: asis` is passed. 

https://github.com/quarto-dev/quarto-cli/blob/0a8a3c6bb5ec09809c9c67c77cdf2fbffd422bd6/src/core/jupyter/jupyter.ts#L1470-L1491

with `asis` being
https://github.com/quarto-dev/quarto-cli/blob/0a8a3c6bb5ec09809c9c67c77cdf2fbffd422bd6/src/core/jupyter/jupyter.ts#L1252-L1258

This is something I already wondered in https://github.com/quarto-dev/quarto-cli/discussions/6874 but did not have time to look into it. I think it is worth discussing to understand what would explain the difference between **Jupyter** and **knitr** here. 

Maybe something is missing in **knitr** output or something needs to be adapted for Quarto.

**Comment by cderv** [comment]
Reopening only to mark it as discussion

**Comment by mcanouil** [comment]
Thanks @cderv as usual for going in depth into these subjects ;)

**Comment by AlbertRapp** [comment]
Thank you @cderv for this thorough comparison. You wrote in your headline

> `output: asis` does not always add a `div`

And this is the exact thing that confused me. I have used `output:asis` multiple times already and never had much trouble. Probably because I always had use cases where `output:asis` doesn't add a `div`. Hence, I was perplexed as to why my understanding of `output:asis` seems to be faulty.

**Comment by AlbertRapp** [comment]
While working with `{htmltools}` and `output: asis` some more, I realized that the output is not even solely surrounded by another container. The HTML code itself gets modified. See example in #7614 .

**Comment by cscheid** [comment]
@AlbertRapp FYI We have a fix incoming for this which requires some documentation changes, but it will strip the `cell-output-display` div from `output: asis`. (With that said, I think that what you're seeing in #7614 is unrelated.)

**Comment by AlbertRapp** [comment]
Glad to hear that. Yeah, I only mentioned this here since both this thread and #7614 use `output: asis` and the output is still subject to change. That's the similarity. Probably still different root causes 😀
